Reorganize the tree

sauclovian-g opened the pull request.

sauclovian-g commented:
This is quite preliminary still; the commits aren't stable yet. Also, the reorg sits on top of changes to the way GitRev.hs happens, because that seemed reasonable at the time, and that is going to get merged separately before anything happens with this branch. And it isn't fully ready yet and will currently break the docker containers.

The real reorg part starts with "cryptol-saw-core shouldn't (and doesn't) depend on saw-core-*."

I would not bother wading through the individual commits, though, except maybe the "tidy" ones up front; things have been moved around in smallish chunks both to avoid provoking git and to avoid having points in the sequence where the tree was any more broken than unavoidably necessary. Also in some places the moves are in two steps because the plan changed and I wanted to be able to revert to the old plan if necessary.

Rather, I would look through the resulting tree to see how it looks.

...it also seems that while it builds for me, it doesn't build for the CI...

sauclovian-g commented:
ah, it's because the gitrev stuff breaks the CI too at this point. no matter, will get attended to.

RyanGlScott commented:
Here are my initial reactions:

* I think `saw-central` is a perfectly fine name.
* Regarding the fact that the top-level file is named `saw-script.cabal`: what if we renamed it to `saw.cabal` and renamed what is currently just called "`library"` in the `.cabal` file to `library saw-script`? This would avoid the issue of having a sublibrary with the same name as the package, and it would make the name of the `.cabal` file a bit closer in spirit to what it is trying to do (namely, organize everything related to SAW).
* I am not sure if I like having `rme` and `verif-viewer` be part of `saw-script.cabal`, as neither of them depend on any "SAW-like" sublibraries. I would favor just keeping them as separately `.cabal` projects.
* This renames the `heapster-saw` library to `heapster`, but it keeps the sources under a subdirectory that is still named `heapster-saw/`. I am not fond of this—I'd rather have the library and its source directory have the same name. I don't have a particularly strong opinion on whether that name should be `heapster-saw` or `heapster`, however.
* What is the rationale for renaming the `saw-remote-api` library to `saw-server`? The `saw-remote-api` executable still has the same name, so I wonder why only the library was renamed.
* I take it that the `saw-tools/` directory is meant to contain all of the executables that aren't `saw` and `saw-remote-api`? If so, it would be worth stating as much in a small `saw-tools/README.md` file.
* I am not sure if I understand what goes in `otherTests/`. It includes things like the `saw-core` test suite, but it excludes things like `heapster`'s `proverTests`. Is there a reason for this?

Since this is still a draft, I haven't looked too deeply into anything else (e.g., CI). Please let me know when this needs a more thorough review.

sauclovian-g commented:
Thanks :-)

Initial thoughts:
- ```saw-central``` it is then, unless someone else comes along with a clearly better idea
- Then it has the same name as the ```saw``` executable. But I guess that's ok. I'll do this, I think it's an improvement.
- serif-viewer is really part of saw so I think it should be in the umbrella. Also, having it be its own project adds a lot of administrative complexity. rme on the other hand is its own thing and it's reasonable to exclude it.
- My inclination would be to rename the directory (it is possible that I originally intended to and it got left off by accident, too)... never been clear on what the ```-saw``` was supposed to mean.
- Re ```saw-server``` the idea was to split off the library. Changing the name of the executable would be a bad idea, and the module namespace was already called ```SAWServer``` so this was the path of least resistance. The library doesn't _need_ to be split off but it's definitely tidier if each executable is just a Main.hs.
- Re ```saw-tools```, yes, and that's a good idea.
- ```otherTests``` is all the random additional test suites I found, other than heapster's, which I've left where it is for heapster reasons. If you think that should get moved into ```otherTests``` anyway, can easily do it. (And are there any others I missed? I think at this point everything else is either ```intTests``` or ```s2nTests``` or in a subpacakge.)

Re real review, don't worry, I'll ping for actual reviews once it's something like ready.

RyanGlScott commented:
> * My inclination would be to rename the directory (it is possible that I originally intended to and it got left off by accident, too)... never been clear on what the `-saw` was supposed to mean.

I think the intent was to separate out the Haskell-specific parts of the Heapster code base (which are baked into the `saw` tool, hence `heapster-saw`) from the Coq-specific parts (which are sprinkled throughout the Coq support libraries in `saw-core-coq`). It's not a terribly meaningful distinction, in my view. I'd be fine with just renaming it to `heapster`.

> * `otherTests` is all the random additional test suites I found, other than heapster's, which I've left where it is for heapster reasons. If you think that should get moved into `otherTests` anyway, can easily do it.

Heapster's testing situation is a bit unusual in that it is split between Haskell-based tests (i.e., `proverTests`) and Coq-based tests (which are found in the [`examples/mr_solver`](https://github.com/GaloisInc/saw-script/tree/8456cf84135c6c7762e0b9570bb86cacb769ca13/examples/mr_solver) and [`heapster-saw/examples`](https://github.com/GaloisInc/saw-script/tree/8456cf84135c6c7762e0b9570bb86cacb769ca13/heapster-saw/examples) directories). I suppose there is some merit in trying to keep all of the Heapster-related tests under `heapster-saw/` (or `heapster/`). Alternatively, we could put the Haskell-based tests under `otherTests` and the Coq-based tests somewhere else. I don't have a particularly strong opinion, although it would be good to document whatever choice we do make.

> (And are there any others I missed? I think at this point everything else is either `intTests` or `s2nTests` or in a subpacakge.)

I think you've covered everything.

sauclovian-g commented:
I have done the following:
- renamed `saw-script.cabal` to `saw.cabal`
- renamed `heapster-saw` to `heapster`
- dropped the rme changes (but I kept the changeset locally in case we decide we want it back)
- added a saw-tools readme file
- fixed some other stuff I found that was wrong
- rebased on new gitrev stuff so the CI and Docker images should work again

I'm not sure whether to move `proverTests` into `otherTests`. Since on examination it's actually just a single source file, I'm now sort of inclined to go ahead. (But if I do, should it get renamed to `heapsterTests`?)

...hmm every time I force-push it's going to generate a ton of scroll

sauclovian-g commented:
At long last, the docker build _and_ the remote API tests finally succeeded: https://github.com/GaloisInc/saw-script/actions/runs/14392480596

Consequently, I think this is now unbroken, so I'm going to mark it ready for review.

The commits aren't really stable yet, because there's a bunch of squashing still to do. The renames are in batches on purpose for git safety, but a lot of them are done in two stages (because of the question of whether e.g. `saw-central/src` should exist as a separate layer or not) and unless anyone has better ideas for what to do about that, those are going to be condensed.

Plus the goal is to merge #1869, #1827, and #927 (maybe also #2288) before merging here, and because this kind of thing should specifically not end in a merge commit, I'm going to rebase over all of those.

For reviewing, if the overall files-changed view works for you great, if not my suggestion would be to use it to see what goes where (note that this is also documented in doc/developer/reorg-2025.md and if you notice any discrepancies that's a bug) and look at the individual commits that _aren't_ renames. But unfortunately a lot of those are also full of noise (because of the need to adjust module names and imports everywhere) so I'm not sure either approach is going to work that well.

RyanGlScott commented on cabal.project:
```diff
@@ -1,17 +1,9 @@
 packages:
-    saw-script.cabal
-    saw-remote-api
+    rme
+    saw.cabal
```
A minor thing, but I'd prefer to list `saw` as the first package, as it's by far the most important one.

sauclovian-g replied:
I tend to do this kind of thing in dependency order by default, but I don't really care one way or the other.

RyanGlScott commented on saw.cabal:
```diff
@@ -0,0 +1,1136 @@
+
+-- The saw-script library.
+--
+-- This can't be a sublibrary because you can't have a sublibrary with
+-- the same name as the package (why?) but it should not be construed as
+-- bigger or more important than the others.
```
Is this comment obsolete now that we have renamed the `.cabal` file from `saw-script.cabal` to `saw.cabal`?

sauclovian-g replied:
Yes. Removed.

RyanGlScott reviewed (approved):
I have not looked at every line of the diff, but I am quite happy with how this has turned out. Thank you so much for all your persistence in working on this!

sauclovian-g commented:
Growl. Bitten again by not building test programs by default 😿

sauclovian-g commented:
I rebased this over the quickcheck-in-remote-API change (#2293). Current plan is to merge the upcoming Cryptol enum support patches first, rebase over that, then merge this.

sauclovian-g commented:
ok, this version is now probably (at last) final
